pkg/util/resourceexecutor/executor.go

```diff
@@ -0,0 +1,239 @@
+/*
+Copyright 2022 The Koordinator Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package resourceexecutor
+
+import (
+	"sync"
+	"time"
+
+	"k8s.io/klog/v2"
+
+	"github.com/koordinator-sh/koordinator/pkg/tools/cache"
+)
+
+const (
+	defaultExecutorUpdatesBufferSize = 2
+	defaultWorkerNumber              = 2
+)
+
+var _ ResourceUpdateExecutor = &ResourceUpdateExecutorImpl{}
+
+type ResourceUpdateExecutor interface {
+	Update(cacheable bool, resource ResourceUpdater) (updated bool, err error)
+	UpdateBatch(cacheable bool, resources ...ResourceUpdater)
+	// LeveledUpdateBatch is to cacheable update resources by the order of resources' level.
+	// For cgroup interfaces like `cpuset.cpus` and `memory.min`, reconciliation from top to bottom should keep the
+	// upper value larger/broader than the lower. Thus a Leveled updater is implemented as follows:
+	// 1. update batch of cgroup resources group by cgroup interface, i.e. cgroup filename.
+	// 2. update each cgroup resource by the order of layers: firstly update resources from upper to lower by merging
+	//    the new value with old value; then update resources from lower to upper with the new value.
+	LeveledUpdateBatch(cacheable bool, resources [][]ResourceUpdater)
+	Run(stopCh <-chan struct{})
+}
+
+var singleton ResourceUpdateExecutor
+
+type UpdateRequest struct {
+	Updater   ResourceUpdater
+	Cacheable bool
+}
+
+type ResourceUpdateExecutorImpl struct {
+	lock               sync.Mutex // used for leveled update
+	resourceCache      *cache.Cache
+	forceUpdateSeconds int
+	workers            int
+	updates            chan *UpdateRequest
+}
+
+func NewResourceUpdateExecutor() ResourceUpdateExecutor {
+	if singleton == nil {
```

Comment:
in case some modules call the `NewResourceUpdateExecutor()` individually, we may keep the singleton pattern and use `sync.Once` to solve the concurrency problem